Adding Support for Key Pair Authentication

rymancl opened the pull request:
We have a strict requirement for using private key authentication when running Snowflake scripts in our automation pipelines. This PR for **version 2.5.0** is adding support for encrypted key pair authentication based on the [docs for the Python connector](https://docs.snowflake.com/en/user-guide/python-connector-example.html#using-key-pair-authentication).

Note that I have prioritized password authentication due to the fact it was in the script first. If this logic should be changed, let me know.

sfc-gh-jhansen commented:
This is great @rymancl, and will help a lot of people. Thank you for taking the time to give back. One quick question, what do you think about renaming the new environment variable names to match the existing pattern? With the (legacy) exception of the password using "SNOWSQL_" all of the environment variables begin with "SNOWFLAKE_". Would be nice to be consistent with these new ones, unless I'm missing something. Thanks again!

rymancl commented:
@sfc-gh-jhansen Heh, it's funny you mention that. I actually had renamed it like you mentioned, but changed it back because I didn't want to introduce breaking changes. I'd love to get rid of the "SNOWSQL_" prefix, but I'll defer to your judgement on it.

Just let me know either way, I'm happy to change it if you'd like. 

Thanks!

sfc-gh-jhansen commented:
Yeah, the SNOWSQL_PWD name is legacy, from the first version of snowchange. I believe it was simply to fall inline with our SnowSQL CLI tool which uses that environment variable name. I'm inclined to leave it for now, or at least allow for both (SNOWSQL_PWD and SNOWFLAKE_PASSWORD) and mark the old one as deprecated in the docs for now. I'm good with either one, just don't want to remove it without giving folks the heads up.

rymancl commented:
@sfc-gh-jhansen I liked your idea of having support for both, so I've implemented and tested that. I added a deprecation notice for `SNOWSQL_PWD` in both the code and the README. It'll print a message similar to this when that environment variable is set:

>snowchange.py:165: DeprecationWarning: The SNOWSQL_PWD environment variable is deprecated and will be removed in a later version of snowchange. Please use SNOWFLAKE_PASSWORD instead.
  warnings.warn("The SNOWSQL_PWD environment variable is deprecated and will be removed in a later version of snowchange. Please use SNOWFLAKE_PASSWORD instead.", DeprecationWarning)

Let me know what you think.

Thanks!

sfc-gh-jhansen commented:
Looking better @rymancl, thanks so much! But I noticed that the new environment variables still aren't following the "SNOWFLAKE_" naming convention, which was my original concern/suggestion. Looks like they're still named "PRIVATE_KEY_PATH" and "PRIVATE_KEY_PASSPHRASE". Can you please update those so that all environment variables follow the same naming conventions? Thanks again.

rymancl commented:
@sfc-gh-jhansen Fixed!

sfc-gh-jhansen commented:
Awesome, thanks @rymancl, just merged.
